src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java

```diff
@@ -529,39 +542,62 @@ public void setUrlPrefix(String urlPrefix) {
     }
 
     /**
-     * Get the name of the ctags program in use
-     *
-     * @return the name of the ctags program in use
+     * Gets the name of the ctags program to use: either the last value passed
+     * successfully to {@link #setCtags(java.lang.String)}, or
+     * {@link Configuration#getCtags()}, or the system property for
+     * {@code "org.opensolaris.opengrok.analysis.Ctags"}, or "ctags" as a
+     * default.
+     * @return a defined value
      */
     public String getCtags() {
-        return threadConfig.get().getCtags();
+        String value;
+        return ctags != null ? ctags : (value =
+            threadConfig.get().getCtags()) != null ? value :
+            System.getProperty("org.opensolaris.opengrok.analysis.Ctags",
+            "ctags");
     }
 
     /**
-     * Specify the CTags program to use
+     * Sets the name of the ctags program to use, or resets to use the fallbacks
+     * documented for {@link #getCtags()}.
+     * <p>
+     * N.b. the value is not mediated to {@link Configuration}.
      *
-     * @param ctags the ctags program to use
+     * @param ctags a defined value or {@code null} to reset to use the
+     * {@link Configuration#getCtags()} fallbacks
+     * @see #getCtags()
      */
     public void setCtags(String ctags) {
-        threadConfig.get().setCtags(ctags);
```

Comment:
What's the logic behind this change ? Also I wonder what will be the effect w.r.t. `Configuration` serialization - will the XML form still contain custom ctags value ?

Reply:
(There are no non-test users of `RuntimeEnvironment setCtags(String)`.)

Right now, `Configuration` has a non-static default value for `ctags` (from the env var, `org.opensolaris.opengrok.analysis.Ctags`) so `OpenGrok` has the non-intuitive behavior that if you specify `-c,--ctags ...` on the command-line when you have the environment variable set, then no value will be written to `configuration.xml` because the serializer determines that `Configuration` matches a default-constructed instance.

This just moves the reading of the env var to `RuntimeEnvironment` as a runtime concern vs a configuration concern and so that `configuration.xml` will always have a defined element when you specify `-c,--ctags ...` on the command-line.

Reply:
Makes sense. Will this make code like

```
private static final String CTAGS_PROP =
60          "org.opensolaris.opengrok.analysis.Ctags";
...

env.setCtags(System.getProperty(CTAGS_PROP, "ctags"));
```

in `JavaAnalyzerFactoryTest.java` redundant ?

Reply:
Yes, I think so. Let me push a cleanup of those — unless you think to wait.

Reply:
Go ahead, I think it makes to have the cleanup done in the same PR.